workflow/sync/multi_throttler.go

```diff
@@ -0,0 +1,243 @@
+package sync
+
+import (
+	"container/heap"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/tools/cache"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+)
+
+//go:generate mockery --name=Throttler
+
+// Throttler allows the controller to limit number of items it is processing in parallel.
+// Items are processed in priority order, and one processing starts, other items (including higher-priority items)
+// will be kept pending until the processing is complete.
+// Implementations should be idempotent.
+type Throttler interface {
+	Init(wfs []wfv1.Workflow) error
+	Add(key Key, priority int32, creationTime time.Time)
+	// Admit returns if the item should be processed.
+	Admit(key Key) bool
+	// Remove notifies throttler that item processing is no longer needed
+	Remove(key Key)
+}
+
+type Key = string
+type QueueFunc func(Key)
+
+// NewMultiThrottler creates a new multi throttler for throttling both namespace and global parallelism, a parallelism value of zero disables throttling
+func NewMultiThrottler(parallelism int, namespaceParallelism map[string]int, namespaceParallelismDefault int, queue QueueFunc) Throttler {
+	return &multiThrottler{
+		queue:                       queue,
+		namespaceParallelism:        namespaceParallelism,
```

Comment:
I understand why you have this, but as it is unusable by workflow users and I had imagined it might be a callback (to copy the pattern set by `queueFunc`) I wonder if we should keep in in this PR?

I'm OK with leaving it in, it's just a bit odd in this PR.

Reply:
I could hide the implementation detail (map as an argument) from this PR in that case. I mean to immediately follow up with a PR that passes in a config containing limits for each namespace.

Reply:
This will presumably not be a constant map (as in parallelism for a namespace will be modifyable on the fly without restarting the controller). The expectation was an annotation on the namespace or a configmap in the namespace.

Reply:
No it will not be, I will follow up with a change to make it dynamic.